src/plugins/intel_cpu/src/emitters/snippets/x64/jit_brgemm_emitter.cpp

```diff
@@ -5,23 +5,150 @@
+        CPUKernelExecutor<BrgemmKernelConfig, brgemm_kernel_t>(kernel_cache, config) {
+    if (config->is_complete())
+        update_kernel();
+}
+
+
+std::shared_ptr<brgemm_kernel_t> BrgemmKernelExecutor::compile_kernel(const std::shared_ptr<BrgemmKernelConfig>& config) const {
+    OV_CPU_JIT_EMITTER_ASSERT(config, "Invalid config provided for BrgemmKernelDesc::compile_kernel");
+    cpu::x64::brgemm_t desc;
+    auto status = brgemm_desc_init(&desc, config->isa, cpu::x64::brgemm_strd,
+                                   config->dt_in0, config->dt_in1,
+                                   false, false, cpu::x64::brgemm_row_major, 1.f,
+                                   config->beta,
+                                   config->LDA, config->LDB, config->LDC,
+                                   config->M, config->N, config->K, nullptr);
+
+    OV_CPU_JIT_EMITTER_ASSERT(status == dnnl_success, "Cannot initialize brgemm descriptor due to invalid params");
+    if (config->is_with_amx) {
+        status = brgemm_init_tiles(desc, config->palette);
+        OV_CPU_JIT_EMITTER_ASSERT(status == dnnl_success, "Cannot initialize brgemm tiles due to invalid params");
+    }
+
+    cpu::x64::brgemm_kernel_t* kernel_ = nullptr;
+    status = brgemm_kernel_create(&kernel_, desc);
+    OV_CPU_JIT_EMITTER_ASSERT(status == dnnl_success, "Cannot create brgemm kernel due to invalid params");
+    return std::unique_ptr<brgemm_kernel_t>(kernel_);
+}
+
+void BrgemmKernelExecutor::execute(const BrgemmKernelExecutor* desc, call_args* args) {
+    const auto& config = desc->m_config;
+    if (config->is_with_amx) {
+        const auto& amx_tile_config = args->amx_tile_config;
+        if (config->M != amx_tile_config->M || config->K != amx_tile_config->K || config->N != amx_tile_config->N) {
+            amx_tile_config->M = config->M;
+            amx_tile_config->K = config->K;
+            amx_tile_config->N = config->N;
+            cpu::x64::amx_tile_configure(config->palette);
+        }
+    }
+
+    cpu::x64::brgemm_kernel_params_t brgemm_p;
+
+    brgemm_p.batch = nullptr;  // default value
+    brgemm_p.ptr_A = args->A;
+    brgemm_p.ptr_B = args->B;
+    brgemm_p.ptr_C = args->C;
+    brgemm_p.ptr_D = args->C;
+    brgemm_p.ptr_buf = args->scratch;
+    brgemm_p.ptr_bias = nullptr;
+    brgemm_p.do_post_ops = static_cast<size_t>(config->is_with_comp);
+    brgemm_p.do_apply_comp = static_cast<size_t>(config->is_with_comp);
+    brgemm_p.skip_accm = 0;
+    brgemm_p.BS = 1;  // default value
+    OV_CPU_JIT_EMITTER_ASSERT(desc->m_kernel, "has nullptr kernel");
+    (*desc->m_kernel)(&brgemm_p);
+}
+
+void BrgemmKernelExecutor::update_kernel_config(size_t M, size_t N, size_t K, size_t LDA, size_t LDB, size_t LDC) {
+    OV_CPU_JIT_EMITTER_ASSERT(m_config, "update_kernel_config is called for empty kernel config");
+#define CAST(X) m_config->X = DIM_CAST(X)
+    CAST(M); CAST(N); CAST(K);
+    CAST(LDA); CAST(LDB); CAST(LDC);
+#undef CAST
+    update_kernel();
```

Comment:
The logic behind this design is very straightforward: `KernelExecutor` owns config and kernel and it always maintains 1-1 correspondence between them. That's why config update have to trigger kernel automatically.

Of cause, you can shift the burden of maintaining the to the user (e.g. remove `update_kernel` from `update_kernel_config`) but it is a error-prone design, since the user can forget to do it. And even if they doesn't, they always need to call `update_state()` after every `update_kernel_config(..)`. So why don't do it automatically?

Reply:
Discussed offline. Decided to rename the method to `update`, and add a description